src/bidiMapper/modules/context/NavigationTracker.ts

```diff
@@ -72,210 +102,365 @@ export class NavigationTracker {
     url: string,
     browsingContextId: string,
     eventManager: EventManager,
+    logger?: LoggerFn,
   ) {
     this.#browsingContextId = browsingContextId;
-    this.#url = url;
     this.#eventManager = eventManager;
+    this.#logger = logger;
+
+    this.#isInitialNavigation = true;
+    this.#currentNavigation = new NavigationState(url, browsingContextId);
+  }
+
+  #createNavigation(url: string): NavigationState {
+    const navigation = new NavigationState(url, this.#browsingContextId);
+    this.#setListeners(navigation);
+    return navigation;
   }
 
+  /**
+   * Returns current started ongoing navigation. It can be either a started pending
+   * navigation, or one is already navigated.
+   */
   get currentNavigationId() {
+    if (this.#pendingNavigation?.loaderId !== undefined) {
+      return this.#pendingNavigation.navigationId;
+    }
+
     return this.#currentNavigation.navigationId;
   }
 
-  get initialNavigation(): boolean {
-    return this.#initialNavigation;
+  /**
+   * Flags if the current navigation relates to the initial to `about:blank` navigation.
+   */
+  get isInitialNavigation(): boolean {
+    return this.#isInitialNavigation;
   }
 
-  get pendingCommandNavigation(): Deferred<void> | undefined {
-    return this.#pendingCommandNavigation;
+  /**
+   * Url of the last navigated navigation.
+   */
+  get url(): string {
+    return this.#currentNavigation.url;
   }
 
-  get url(): string {
-    return this.#url;
+  /**
+   * Creates a pending navigation e.g. when navigation command is called. Required to
+   * provide navigation id before the actual navigation is started. It will be used when
+   * navigation started. Can be aborted, failed, fragment navigated, or became a current
+   * navigation.
+   */
+  createPendingNavigation(url: string): NavigationState {
+    this.#logger?.(LogType.debug, 'createCommandNavigation');
+    if (!urlMatchesAboutBlank(url)) {
+      this.#isInitialNavigation = false;
+    }
+
+    this.#pendingNavigation?.finished.resolve(
+      new NavigationResult(
+        NavigationEventName.NavigationAborted,
+        'navigation canceled by concurrent navigation',
+      ),
+    );
+    const navigation = this.#createNavigation(url);
+    this.#pendingNavigation = navigation;
+    return navigation;
+  }
+
+  #setListeners(navigation: NavigationState) {
+    void navigation.started
```

Comment:
so the way of dispatching events is not working as expected. Using Deferred.then creates a new microtask that makes events emitted to be emitted at a different relative order to the events scheduled when the deferred is resolved. I believe this is the reason for the failing data URL tests. The correct way to handle event would be to register the event at the same time as the deferred is resolved.

Reply:
in the case of the data URLs it allows network storage to dispatch the beforeRequest event before navigationStarted because network storage does not delay the event registration